Proposal: Searching for implementations with keys

NickNeck opened the pull request:
Hi, Sascha

my suggestion would be to extend the `config_key` option so that a list is also accepted.
That would be usefull for `behaviour`s with additional configurations.
Example:
```elixir
config :my_app, :sys,
  module: MyApp.Sys.Impl
  timeout: 2_000
```
```
defmodule MyApp.Sys do
  use Knigge,
    otp_app: :my_app,
    config_key: [:sys, :module]
  ...
end
```
What do you think?

Kind regards, Marcus

p.s. Nice work and I love the name of the lib.

NickNeck committed:
Add searching for implementations with keys

NickNeck committed:
Run mix format

sascha-wolf commented on lib/knigge/implementation.ex:
```diff
@@ -9,6 +9,12 @@ defmodule Knigge.Implementation do
 
   alias Knigge.Options
 
+  def fetch!(%Options{implementation: {:config, otp_app, [key | keys]}}) do
+    otp_app
+    |> Application.fetch_env!(key)
+    |> get_in(keys)
+  end
+
```
Sadly, this doesn't really play nice with the `default` changes from #18.

It would also fail quietly when `key` exists but `get_in` returns `nil` for `keys`.

Instead I'd would expect to receive an error along the lines of:

> Unable to fetch implementation for `MyFacade` from config of app `<otp_app>` under path `<keys>`.

The message could be improved upon but I think you catch my drift?

sascha-wolf commented on lib/knigge/options.ex:
```diff
@@ -305,7 +305,7 @@ defmodule Knigge.Options do
     implementation: :module,
     otp_app: :atom,
-    config_key: :atom,
+    config_key: :keys,
```
Instead of introducing `:keys` I'd argue it's a good opportunity to introduce an "or" mechanism:

```elixir
config_key: [:atom, {:list_of, :atom}]
```

Where `valid_value?/2` would look like this:

```elixir
defp valid_value?({:list_of, type}, value),
  do: is_list(value) and Enum.all?(value, &valid_value?(type, &1))

defp valid_value?(list, value) when is_list(list),
  do: Enum.any?(list, &valid_value?(&1, value))
```

In addition the `transform` function could then be extended to map `config_key` to always be a list (by using `List.wrap/1`).

Which in turn would simplify the implementation in `Implementation.fetch!/1` as there no need for multiple clauses anymore (one for single atom, one for list).

NickNeck committed:
Update option transformation

NickNeck committed:
Update search with config_key

NickNeck committed:
Fix test

NickNeck committed:
Fix test

NickNeck committed:
Fix test

escobera commented:
Hey @sascha-wolf what do you think is needed to merge this? I'm available to help if needed. Could really use this feature =)

sascha-wolf commented on lib/knigge/implementation.ex:
```diff
@@ -9,11 +9,13 @@ defmodule Knigge.Implementation do
-      Application.fetch_env!(otp_app, key)
-    else
-      Application.get_env(otp_app, key, default)
+  def fetch!(%Options{implementation: {:config, otp_app, [key | keys]}, default: default}) do
+    with nil <- otp_app |> env!(key, default) |> get(keys) do
+      raise ArgumentError,
+        message: """
+        could not fetch application environment #{inspect([key | keys])} \
+        for application #{inspect(otp_app)}\
+        """
```
If it's okay I'd prefer to write it like this:

```suggestion
  def fetch!(%Options{implementation: {:config, otp_app, [key | keys]}, default: default}) do
    otp_app
    |> env!(key, default)
    |> get(keys)
    |> case do
      nil ->
        raise ArgumentError,
          "could not fetch application environment #{inspect([key | keys])} for application #{inspect(otp_app)}"
        
      implementation ->
        implementation
    end
```

sascha-wolf commented on lib/knigge/options.ex:
```diff
@@ -128,8 +128,8 @@ defmodule Knigge.Options do
       |> map_deprecated()
       |> validate!()
-      |> with_defaults()
       |> transform(with_env: env)
+      |> with_defaults()
 
```
What's the train of thought here of swapping these two? This would actually break the `delegate_at_runtime?` default, since the "environment" spec gets resolved to a boolean in `transform/2`.

sascha-wolf commented on lib/knigge/options.ex:
```diff
@@ -331,11 +333,22 @@ defmodule Knigge.Options do
   defp valid_value?(:module, value), do: is_atom(value)
   defp valid_value?(:keyword, value), do: Keyword.keyword?(value)
 
+  defp valid_value?(:keys, value),
+    do: is_atom(value) || (is_list(value) && Enum.all?(value, &is_atom/1))
+
```
I think we can drop this now, right?
```suggestion
```

sascha-wolf commented on lib/knigge/options.ex:
```diff
@@ -352,6 +365,9 @@ defmodule Knigge.Options do
       :keyword ->
         "keyword list"
 
+      [:atom, {:list_of, :atom}] ->
+        "atom or list of atoms"
+
```
Technically we could make this more general purpose but for now this is sufficient.

I'd only suggest to add a little reminder to future us:

```suggestion
      # For now we explicitly match on the "or" option, as soon as we add further 
      # "or"ed options this ought to be refactored to something more general purpose
      [:atom, {:list_of, :atom}] ->
        "atom or list of atoms"

```

sascha-wolf reviewed (dismissed):
Apart from some cosmetic suggestions and one small concern, LGTM 💙💜❤️🧡💛💚

NickNeck committed:
Update lib/knigge/options.ex

Co-authored-by: Sascha Wolf <swolf.dev@gmail.com>

NickNeck committed:
Update .tool-versions

NickNeck committed:
Updates for PR #17

NickNeck committed:
Run mix format

NickNeck committed:
Update ci workflow

NickNeck committed:
Set fail-fast to false

NickNeck committed:
Fix fail-fast

NickNeck committed:
Unlock unused dep earmark

NickNeck commented:
Hi, Sascha

I have made some more changes and incorporated your change requests.

I have also updated the `deps` and the `ci.yml`.

Because `use Mix.Config` is deprecated I have changed it to `import Config`. This causes the failing tests for Elixir versions smaller as `1.9.0`.

Let me know if the PR is too big and how we want handle the failing tests.

Have a nice day.

sascha-wolf commented on test/behaviour/with_otp_app_key_test.exs:
```diff
@@ -29,6 +56,22 @@ defmodule Behaviour.WithOtpAppKey do
+
+    # Should not raise
+    behaviour =
+      defmodule_salted WorkingBehaviour do
+        use Knigge,
+          otp_app: :knigge,
+          config_key: [:sys, :module]
+      end
```
Reading this test, I think it'd be beneficial for readability to deprecate `config_key` in favor of `config_path`. The `Options` module has a `map_deprecated` function ([link](https://github.com/sascha-wolf/knigge/blob/e0886d345c7ee9af5be716a33b78d4f358cc999b/lib/knigge/options.ex#L170-L175)) which we can use to map `config_key` to `config_path` while also printing a deprecation warning.

What do you think?

NickNeck replied:
Good idea. I will add this.

NickNeck commented:
I just update the `ci.yml` to test against the latest versions. I used my little [`git_hub_actions`](https://github.com/hrzndhrn/git_hub_actions) tool to make things simpler for me. I will make a separate PR and then you can take look if it is okay for you.

SukhikhN commented:
Hello @NickNeck and @sascha-wolf!
Would you be so kind as to bring this PR to a release?

Now I fetch an implementation module from nested configuration key and pass it to `Knigge` directly with the `:implementation` option, but with this approach I can't change it in runtime.

sascha-wolf commented:
> Hello @NickNeck and @sascha-wolf! Would you be so kind as to bring this PR to a release?

I'll try to find some time this week. Life has not been kind lately.
